addons/website_forum/tests/test_controller.py

```diff
@@ -0,0 +1,54 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import json
+from odoo.tests.common import HttpCase, new_test_user
+
+
+class TestController(HttpCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        portal_user = new_test_user(cls.env, login='portal_user', groups='base.group_portal')
+        cls.portal = portal_user.login
+        admin_user = new_test_user(cls.env, login='admin_user', groups='base.group_user,base.group_system')
+        cls.admin = admin_user.login
+        cls.headers = {"Content-Type": "application/json"}
+        cls.pixel = 'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAIAAACQd1PeAAAADElEQVR4nGNgYGAAAAAEAAH2FzhVAAAAAElFTkSuQmCC'
+
+    def _build_payload(self, params=None):
+        """
+        Helper to properly build jsonrpc payload
+        """
+        return {
+            "jsonrpc": "2.0",
+            "method": "call",
+            "id": 0,
+            "params": params or {},
+        }
+
+    def test_portal_attachment_01(self):
+        self.authenticate(self.portal, self.portal)
+        payload = self._build_payload({"name": "pixel", "data": self.pixel, "is_image": True, "res_model": "forum.post"})
+        response = self.url_open('/web_editor/attachment/add_data', data=json.dumps(payload), headers=self.headers)
+        self.assertEqual(200, response.status_code)
+        attachment = self.env['ir.attachment'].search([('name', '=', 'pixel')])
+        self.assertTrue(attachment)
+
+    def test_admin_attachment_01(self):
+        self.authenticate(self.admin, self.admin)
+        payload = self._build_payload({"name": "pixel", "data": self.pixel, "is_image": True, "res_model": "forum.post"})
+        response = self.url_open('/web_editor/attachment/add_data', data=json.dumps(payload), headers=self.headers)
+        self.assertEqual(200, response.status_code)
+        attachment = self.env['ir.attachment'].search([('name', '=', 'pixel')])
+        self.assertTrue(attachment)
+
+    def test_admin_attachment_02(self):
+        attachment = self.env['ir.attachment'].create([{'name': 'test_pixel', 'public': True, 'res_id': False,
+                        'mimetype': 'text/plain', 'res_model': 'forum.post', 'raw': self.pixel, 'website_id': 1}])
+        self.authenticate(self.admin, self.admin)
```

Comment:
```suggestion
        self.assertTrue(attachment)
```

Just merge both test, instead of:
Test 1: Check if create goes well
Test 2: Simulate create and check if fetch goes well

Do: Check if create goes well and check if fetch goes well.

Easier to read + better perfs (no need to spawn a new whole browser(?) / generate a new routing map etc

Reply:
Attachments created from a `forum.post` using the controller are not public so the admin cannot fetch them. That's why I have to simulate a create with public=True